### PR TITLE
Add file properties to rclone executable for windows (fixes #4304)

### LINKED ISSUE #4304: Add windows file properties (version info and icon resources)
It would be nice if rclone.exe on windows would include file properties such as product name, file description, version, as well as an icon.

The most useful would be the version information, then one could use PowerShell to retrieve the numerical version number `(Get-Item rclone.exe).VersionInfo.FileVersion`, and for instance compare it with latest released version by accessing GitHub API. I know I can execute `rclone version`, but then I need to perform string parsing and also I need to run a process instead of just accessing file properties.

Syncthing has implemented this in go already, so I'm thinking it wouldn't be too hard if using it as inspiration.
> On Windows generate a special file which the Go compiler will automatically use when generating Windows binaries to set things like the file icon, version, etc.

(https://github.com/syncthing/syncthing/blob/master/build.go)

Solution seems to be based on executing a command line utility, [GoVersionInfo](https://github.com/josephspurrier/goversioninfo), during the build process.

I could try to implement this myself, if you agree it is a good idea, but I am not proficient in go so may be needing some pointers along the way (for instance I am not sure if this introduces some changes to how the build process must be performed).

Another step up would be to authenticode sign the executable, but that requires a certificate etc so would need a bit more effort - and probably money. See https://github.com/syncthing/syncthing/issues/3420.

### PR DESCRIPTION
**Pull request by albertony**
#### What is the purpose of this change?

Make rclone.exe on windows include file properties such as product name, file description, version, as well as an icon.

The most useful would be the version information, then one could use PowerShell to retrieve the numerical version number `(Get-Item rclone.exe).VersionInfo.FileVersion`, and for instance compare it with latest released version by accessing GitHub API. I know I can execute `rclone version`, but then I need to perform string parsing and also I need to run a process instead of just accessing file properties.

#### Was the change discussed in an issue or in the forum before?

See issue https://github.com/rclone/rclone/issues/4304.

#### Checklist

- [X] I have read the [contribution guidelines](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#submitting-a-pull-request).
- [X] ~I have added tests for all changes in this PR if appropriate.~
- [X] ~I have added documentation for the changes if appropriate.~
- [X] All commit messages are in [house style](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#commit-messages).
- [x] I'm done, this Pull Request is ready for review :-)

### CODE REVIEW

File: bin/cross-compile.go [hunk 1]
```diff
@@ -168,6 +170,86 @@ func buildDebAndRpm(dir, version, goarch string) []string {
...
+			},
+		},
+		"StringFileInfo": M{
+			"FileDescription": "Rsync for cloud storage",
+			"LegalCopyright":  "Nick Craig-Wood http://www.craig-wood.com/nick/",
+			"FileVersion":     version,
+			"ProductVersion":  version,
+			"ProductName":     "Rclone",
+		},
+		//"IconPath": "assets/logo.ico",
```
[the remark this answers is not in the document]
Comment (albertony) [reply bin/cross-compile.go:198]: Yes, I will try to get the icon work as well.

**Comment by albertony** [comment]
Something like this ok?
The FileDescription, CopyRight and CompanyName strings are the most debatable values.
Also a question about the icon: I generated it the logo_symbol_color.svg using www.freeconvert.com/svg-to-ico. The result is nice, I think, but it *could* be considered a bit large (git repo bloat, executable bloat), as it is includes images for all possible sizes resulting in 177KB.

image

```
PS C:\Temp> (Get-Item .\rclone.exe).VersionInfo | fl *

FileVersionRaw     : 9.8.7.0
ProductVersionRaw  : 9.8.7.0
Comments           :
CompanyName        : Nick Craig-Wood http://www.craig-wood.com/nick/
FileBuildPart      : 7
FileDescription    : Rsync for cloud storage
FileMajorPart      : 9
FileMinorPart      : 8
FileName           : C:\Temp\rclone.exe
FilePrivatePart    : 0
FileVersion        : 9.8.7
InternalName       : rclone
IsDebug            : False
IsPatched          : False
IsPrivateBuild     : False
IsPreRelease       : False
IsSpecialBuild     : False
Language           : Language Neutral
LegalCopyright     : Nick Craig-Wood http://www.craig-wood.com/nick/
LegalTrademarks    :
OriginalFilename   : rclone.exe
PrivateBuild       :
ProductBuildPart   : 7
ProductMajorPart   : 9
ProductMinorPart   : 8
ProductName        : Rclone
ProductPrivatePart : 0
ProductVersion     : 9.8.7
SpecialBuild       :
```

**Comment by albertony** [comment]
> So you'd want to run the goversioninfo tool just once, probably in this function
> ...
> func compile(version string) 

Seems the compile operations are running in parallel? To avoid issues related to that, I moved the goversioninfo execution to compileArch, and made sure it produces .syso files with goos_goarch suffix. This avoids a concurrent linux build picking them up. It also avoids concurrent windows/386 and windows/amd64 writing to the same files.

Now goversioninfo is executed once for windows/386 and once for windows/amd64 build. This is probably not necessary (I set an option to mark coff format 64-bit, but seems it does not make any difference to the result in this case). But to avoid it, I think we would have to check all filters in compile() and detect if at least one windows build, and then generate it once before starting the builds?

**Comment by albertony** [comment]
The hack:
`cd .. && go install -mod=mod github.com/josephspurrier/goversioninfo/cmd/goversioninfo`

fails on the build/linux check:
> build flag -mod=mod only valid when using modules

This seems to be due to `GO111MODULE=off` in that check.


Its confusing with all this go modules and vendoring stuf, but I think there are two options:

Forcing GO111MODULE=off always for this command, and use old GOPATH behavior (downloads into $GOPATH/src):

`GO111MODULE=off go get -d github.com/josephspurrier/goversioninfo && go install github.com/josephspurrier/goversioninfo/cmd/goversioninfo`

Forcing GO111MODULE=on always, but move out of go.mod location to avoid it being updated, and with `-mod=mod` to force use of the module cache even if there is a vendor directory present (downloads into $GOPATH/pkg/mod):

`GO111MODULE=on cd .. && go install -mod=mod github.com/josephspurrier/goversioninfo/cmd/goversioninfo`

I will try the last one now, to see if checks go through. If both would work, do you have any preferences?

**Comment by ncw** [comment]
> Something like this ok?

Very nice!

> The FileDescription, CopyRight and CompanyName strings are the most debatable values.

> CompanyName        : Nick Craig-Wood http://www.craig-wood.com/nick/

I'd put `https://rclone.org` here I think

> FileDescription    : Rsync for cloud storage

OK!

> LegalCopyright     : Nick Craig-Wood http://www.craig-wood.com/nick/

That should be "The Rclone Authors" I think (and I should change the one in the LICENSE file!)

> The result is nice, I think, but it _could_ be considered a bit large (git repo bloat, executable bloat), as it includes images for all possible sizes resulting in 177KB.

Presumably the executable gets 177k bigger too? Not the end of the world but something! Does Windows scale the images? If so then we should just put 1 big one in and let Windows scale it.

> Seems the compile operations are running in parallel? To avoid issues related to that, I moved the goversioninfo execution to compileArch, and made sure it produces .syso files with goos_goarch suffix. This avoids a concurrent linux build picking them up. It also avoids concurrent windows/386 and windows/amd64 writing to the same files

Yes the compile operations run in parallel. My comment assumed that you'd only need to run it once, but maybe not I don't know!

> Finally got the non-module goversioninfo package build through all checks!
> Solutions was to do the following (in the Makefile):
> 
> `GO111MODULE=off go get github.com/josephspurrier/goversioninfo/cmd/goversioninfo`

I wasn't sure that would work but good - glad you've found a workaround after a lot of trial and error!

**Comment by albertony** [comment]
Like this then:

image

```
OriginalFilename  : rclone.exe
FileDescription   : Rsync for cloud storage
ProductName       : Rclone
Comments          :
CompanyName       : https://rclone.org
FileName          : C:\Dev\Projects\albertony@GitHub\rclone\build\rclone-v9.8.7-windows-amd64\rclone.exe
FileVersion       : 9.8.7
ProductVersion    : 9.8.7
IsDebug           : False
IsPatched         : False
IsPreRelease      : False
IsPrivateBuild    : False
IsSpecialBuild    : False
Language          : Language Neutral
LegalCopyright    : The Rclone Authors
LegalTrademarks   :
PrivateBuild      :
SpecialBuild      :
FileVersionRaw    : 9.8.7.0
ProductVersionRaw : 9.8.7.0
```


> Presumably the executable gets 177k bigger too?

Yes.

> Does Windows scale the images?

It scales down, only, and not always with great results.

If only 48px version is included the result when showing the "Extra large icons" layout in Windows Explorer (which normally shows 256px icons) is the right icon below:

image

It seems to be somewhat of a standard for application icons to include sizes 256, 48, 32 and 16 px, while 180, 128, 96, 72, 64, 24 are optionaly more exotic intermediates. The originally generated ico file from svg contained all of these, except 180px, resulting in 177KB sized file. If I remove all other than the main sizes (256, 48, 32 and 16 px) the file size reduces to 30KB. Removing more sizes can reduce it further (15KB with only 256px, 19KB with 256 and 32px), but with noticably worse quality.

image

So I conclude that a 30KB ico file with 256, 48, 32 and 16 px sizes in it is a good compromise.

PS: All images have 32-bit color, the 256px version is in PNG format and the others are in BMP. There are more aspects to icon presentation in Windows such as dpi scaling, and to icon file content such as color depth, bmp vs png etc, but I hope its not necessary to dive deeper in to that for now.

**Comment by albertony** [comment]
> I think this looks ready to merge now. Do you agree? I just want to make sure you are happy with the image you've uploaded.

Yes :)

**Comment by ncw** [comment]
Merged :-)

Thank you very much - this is a nice change!

**Comment by albertony** [comment]
Should this not be included in the latest beta builds, from https://beta.rclone.org/v1.52.1-056-g6a2b7b97-beta/ on? It is not..

**Comment by ncw** [comment]
The windows binaries are there now. Are they not working with the new info?

**Comment by albertony** [comment]
The info is not there.

**Comment by albertony** [comment]
Build logs are not published somewhere? It should print "Warning: Windows binaries will not have file information embedded" if the resource generation, including execution of goversioninfo, fails.

**Comment by ncw** [comment]
Oh, foo. The Window build doesn't run `make build_dep`

https://github.com/rclone/rclone/blob/5ca82e2f05eb55eef7d699152edafc6bfa659cd6/.github/workflows/build.yml#L183-L188

And this showing check is not set

https://github.com/rclone/rclone/blob/5ca82e2f05eb55eef7d699152edafc6bfa659cd6/.github/workflows/build.yml#L45-L53

I'll have a think about the best way of fixing this

**Comment by ncw** [comment]
I think this is really easy!

```patch
diff --git a/Makefile b/Makefile
index 25dc499b9..f80be30c1 100644
--- a/Makefile
+++ b/Makefile
@@ -85,12 +85,12 @@ check:      rclone
 # Get the build dependencies
 build_dep:
        go run bin/get-github-release.go -extract golangci-lint golangci/golangci-lint 'golangci-lint-.*\.tar\.gz'
-       GO111MODULE=off go get github.com/josephspurrier/goversioninfo/cmd/goversioninfo
 
 # Get the release dependencies
 release_dep:
        go run bin/get-github-release.go -extract nfpm goreleaser/nfpm 'nfpm_.*_Linux_x86_64.tar.gz'
        go run bin/get-github-release.go -extract github-release aktau/github-release 'linux-amd64-github-release.tar.bz2'
+       GO111MODULE=off go get github.com/josephspurrier/goversioninfo/cmd/goversioninfo
 
 # Update dependencies
```

I'll try that on the CI and if it works I'll merge it!

**Comment by ncw** [comment]
It didn't work... it is proving rather frustrating! Even though goversioninfo is on the path I can't get it to find it and run it...

**Pull request merged**